src/nix-to-python.cc

```diff
@@ -4,8 +4,147 @@
 #include "internal/errors.hh"
 #include "internal/ptr.hh"
 
+#include <iostream>
+
 namespace pythonnix {
 
+PyObject *nixPosToPythonObject(nix::Pos &pos) {
+  PyObject *filename = PyUnicode_FromString(static_cast<std::string>(pos.file).c_str());
+  PyObject *line = PyLong_FromLong(pos.line);
+  PyObject *column = PyLong_FromLong(pos.column);
+  return PyTuple_Pack(3, filename, line, column);
+}
+
+PyObject *nixAttrPathToPythonObject(nix::AttrPath &path) {
+  PyObject* type = PyUnicode_FromString("ATTRPATH");
+
+  PyObject* listObj = PyList_New(path.size());
+  if (!listObj) { return (PyObject *)nullptr; }
+
+  for (std::size_t n = 0; n < path.size(); ++n) {
+    PyObject* expr = nixExprToPythonObject(*(path[n].expr));
+    if (!expr) {
+      Py_DECREF(listObj);
+      return nullptr;
+    }
+    PyObject* symbol = PyUnicode_FromString(static_cast<std::string>(path[n].symbol).c_str());
+    PyObject* tuple = PyTuple_Pack(2, symbol, expr);
+    PyList_SET_ITEM(listObj, n, tuple);
+  }
+  return PyTuple_Pack(2, type, listObj);
+}
+
+PyObject *nixExprToPythonObject(nix::Expr &expression) {
```

Comment:
For large expression it is probably the best to convert the tree lazy to python.

Reply:
I am not the most knowledgeable on writing c-extensions but I agree that lazy evaluation would be helpful. I will be working on this.

Reply:
I have not looked into the nix representation of the ast but it could be an object that is returned from the ast.
This object then has methods that create and return new python objects on demand. If you need some python boiler code for classes it is sometimes easier to just import them from a python module instead of creating them via API: https://github.com/hase-project/hase/blob/master/pt/module.cpp#L67